common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java

```diff
@@ -128,37 +136,58 @@ private void tryCastSpell(SpellUnit a, SpellUnit b, SpellUnit c) {
         List<SpellDirection> spell = Stream.of(a, b, c)
                 .map(x -> (x == SpellUnit.PRIMARY) != isSpellInverted ? SpellDirection.RIGHT : SpellDirection.LEFT)
                 .toList();
-        for (int i = 0; i < spellInProgress.length; ++i) {
-            if (spellInProgress[i] != spell.get(i)) {
-                sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
-                return;
+
+        if (safeCasting.get()) {
+            for (int i = 0; i < spellInProgress.length; ++i) {
+                if (spellInProgress[i] != spell.get(i)) {
+                    sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
+                    return;
+                }
             }
         }
 
         lastSelectedSlot = McUtils.inventory().selected;
         List<SpellDirection> remainder = spell.subList(spellInProgress.length, spell.size());
-        remainder.stream().map(SpellDirection::getSendPacketRunnable).forEach(SPELL_PACKET_QUEUE::add);
+        SPELL_PACKET_QUEUE.addAll(remainder);
     }
 
     @SubscribeEvent
     public void onTick(TickEvent e) {
         if (!Models.WorldState.onWorld()) return;
 
         // Clear spell after the 40 tick timeout period
-        if (spellCountdown > 0 && --spellCountdown <= 0) spellInProgress = SpellDirection.NO_SPELL;
+        if (spellCountdown > 0 && --spellCountdown <= 0) {
+            spellInProgress = SpellDirection.NO_SPELL;
+        }
 
         if (SPELL_PACKET_QUEUE.isEmpty()) return;
         if (--packetCountdown > 0) return;
 
         int currSelectedSlot = McUtils.inventory().selected;
         boolean slotChanged = currSelectedSlot != lastSelectedSlot;
 
-        if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
-        SPELL_PACKET_QUEUE.poll().run();
-        if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(currSelectedSlot));
+        if (slotChanged) {
+            McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
+        }
+
+        // Right clicks need a 1 tick delay between them, left clicks don't
+        boolean didRightClick = false;
+        do {
+            SpellDirection spellDirection = SPELL_PACKET_QUEUE.poll();
+            spellDirection.getSendPacketRunnable().run();
+
+            if (spellDirection == SpellDirection.RIGHT) {
+                didRightClick = true;
+            }
+        } while (!SPELL_PACKET_QUEUE.isEmpty()
```

Comment:
because the mod said that 20 cps is the cap and there are some spells (like archer movement spell) that require only a minimum of right clicks we may need a cps counter

Reply:
Confirmed this.